task_processing/plugins/mesos/execution_framework.py

```diff
@@ -542,12 +545,30 @@ def statusUpdate(self, driver, update):
         if task_id not in self.task_metadata:
             # We assume that a terminal status update has been
             # received for this task already.
-            log.info('Ignoring this status update because a terminal status \
-                update has been receiced for this task already.')
+            log.info('Ignoring this status update because a terminal status'
+                     'update has been receiced for this task already.')
             driver.acknowledgeStatusUpdate(update)
             return
+
         md = self.task_metadata[task_id]
 
+        # If we attempt to accept an offer that has been invalidated by master
+        # for some reason such as offer has been rescinded or we have exceeded
+        # offer_timeout, then we will get TASK_LOST status update back from
+        # mesos master.
+        if task_state == 'TASK_LOST' and \
+                'REASON_INVALID_OFFERS' == str(update.reason):
+            # This task has not been launched. Therefore, we are going to
+            # reenqueue it. We are not propogating any event up to the
+            # application.
+            log.warning('Received TASK_LOST from mesos master because we'
+                        'attemted to accept an invalid offer.')
+            self.task_metadata = self.task_metadata.discard(task_id)
+            self.enqueue_task(md.task_config)
```

Comment:
i'd prefer this to be a separate retrying layer with

    retry_pred = lambda x: str(x.state) == 'TASK_LOST' and str(x.reason) == 'REASON_INVALID_OFFERS'

reasons:

- less code in mesos framework
- library users have better control over the stack
- incoming task_id refactoring is going to make this code tricky - we shouldn't enqueue task with same id but we also shouldn't be mutating task_config.task_id because it won't be recognizable up the stack